pkg/ccl/streamingccl/streamingest/ingest_span_configs_test.go

```diff
@@ -48,163 +39,295 @@ func TestIngestSpanConfigs(t *testing.T) {
 
-		testutils.SucceedsWithin(t, func() error {
-			actualSpanConfigRecords, err := ingestor.accessor.GetSpanConfigRecords(ctx, destTarget)
-			if err != nil {
-				return err
-			}
-			actualPretty := prettyRecords(actualSpanConfigRecords)
-			expectedPretty := prettyRecords(toApply.expectedPersisted)
-			if actualPretty != expectedPretty {
-				return errors.Newf("Actual records %s not equal to expected records %s", actualSpanConfigRecords, toApply.expectedPersisted)
-			}
-			return nil
-		}, time.Second*5)
+		actualSpanConfigRecords, err := ingestor.accessor.GetSpanConfigRecords(ctx, i12DestTarget)
+		require.NoError(t, err)
+		require.Equal(t,
+			replicationtestutils.PrettyRecords(toApply.expectedPersistedUserSpans),
+			replicationtestutils.PrettyRecords(actualSpanConfigRecords),
+			"failed on step %d (0 indexed)", ts)
+	}
+}
+
+// TestIngestSpanConfigsInitialScan tests that the correct span configs are
+// updated after the rangefeed which listens to the span config table completes
+// an initial scan. To do so, the test does the following in a loop:
+//
+// 1. Write some updates to a dummy span config table
+//
+// 2. Open a new span config client on the dummy table, inducing a range feed
+// initial scan
+//
+// 3. Ingest the initial scan updates on the real span config table and validate
+// the expected span configs were updated and deleted.
+func TestIngestSpanConfigsInitialScan(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	toIngestCh := make(chan ingestedRecords)
+	streamingTestKnobs := &sql.StreamingTestingKnobs{
+		RightAfterSpanConfigFlush: initFlushHook(toIngestCh),
+	}
+
+	h, sourceAccessor, sourceTenant, cleanup := replicationtestutils.NewReplicationHelperWithDummySpanConfigTable(ctx, t, streamingTestKnobs)
+	defer cleanup()
+
+	destTenantID := roachpb.MustMakeTenantID(sourceTenant.ID.InternalValue + 1)
+
+	i1Source, i2Source, i12Source, _ := makeTestSpans(sourceTenant.ID)
+	i1Dest, i2Dest, i12Dest, destTenantSplitPoint := makeTestSpans(destTenantID)
+	splitRecord := replicationtestutils.MakeSpanConfigRecord(t, destTenantSplitPoint, 14400)
+
+	type testCase struct {
+		updatesDuringPause         [][]spanconfig.Record
+		deletesDuringPause         [][]spanconfig.Target
+		expectedInitScanUpdates    []spanconfig.Record
+		expectedInitScanDeletes    spanconfig.Targets
+		expectedPersistedUserSpans []spanconfig.Record
+	}
+	execTestCase := func(tc testCase, step int) {
+		if len(tc.deletesDuringPause) == 0 {
+			tc.deletesDuringPause = make([][]spanconfig.Target, len(tc.expectedInitScanUpdates))
+		}
+		for i := range tc.updatesDuringPause {
+			require.NoError(t, sourceAccessor.UpdateSpanConfigRecords(ctx, tc.deletesDuringPause[i], tc.updatesDuringPause[i],
+				hlc.MinTimestamp,
+				hlc.MaxTimestamp), "failed on update %d (index starts at 0)", step)
```

Comment:
If you wanted, you could wrap the execTestCase in t.Run() when called below to avoid having to put the step name in every assertion.

Reply:
aha. good point. done.